src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,215 @@
+						  "RETURNING is not supported in INSERT ... SELECT via coordinator",
+						  NULL, NULL);
+
+		return multiPlan;
+	}
+
+	if (insertSelectQuery->onConflict)
+	{
+		multiPlan->planningError =
+			DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+						  "ON CONFLICT is not supported in INSERT ... SELECT via coordinator",
+						  NULL, NULL);
+	}
+
+	if (PartitionMethod(targetRelationId) == DISTRIBUTE_BY_APPEND)
+	{
+		multiPlan->planningError =
+			DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+						  "INSERT ... SELECT into an append-distributed table is not supported",
+						  NULL, NULL);
+	}
+
+	if (list_length(insertSelectQuery->cteList) > 0)
+	{
+		struct DecrementCTELevelsContext context;
+		List *combinedCTEList = NIL;
+
+		/* prepend top-level CTEs in subquery */
+		combinedCTEList = list_concat(combinedCTEList, insertSelectQuery->cteList);
+		combinedCTEList = list_concat(combinedCTEList, subquery->cteList);
+		subquery->cteList = combinedCTEList;
+
+		context.topLevelCTEList = insertSelectQuery->cteList;
+		context.topLevelsUp = 0;
+
+		/* correct ctelevelsup for range table entries that refer to top-level CTEs */
+		DecrementCTELevelsWalker((Node *) subquery, &context);
+	}
+
+	ReorderInsertSelectTargetLists(insertSelectQuery, insertRte, subqueryRte);
+
+	/* add casts when the SELECT output does not directly match the table */
+	forboth(insertTargetCell, insertSelectQuery->targetList,
+			selectTargetCell, subquery->targetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+		TargetEntry *selectTargetEntry = (TargetEntry *) lfirst(selectTargetCell);
+
+		Var *columnVar = NULL;
+		Oid columnType = InvalidOid;
+		int32 columnTypeMod = 0;
+		Oid selectOutputType = InvalidOid;
+
+		/* indirection is not supported, e.g. INSERT INTO table (composite_column.x) */
+		if (!IsA(insertTargetEntry->expr, Var))
+		{
+			ereport(ERROR, (errmsg("can only handle regular columns in the target "
+								   "list")));
+		}
+
+		columnVar = (Var *) insertTargetEntry->expr;
+		columnType = get_atttype(targetRelationId, columnVar->varattno);
+		columnTypeMod = get_atttypmod(targetRelationId, columnVar->varattno);
+		selectOutputType = columnVar->vartype;
+
+		/*
+		 * If the type in the target list does not match the type of the column,
+		 * we need to cast to the column type. PostgreSQL would do this
+		 * automatically during the insert, but we're passing the the SELECT
+		 * output directly to COPY.
+		 */
+		if (columnType != selectOutputType)
+		{
+			Expr *selectExpression = selectTargetEntry->expr;
+			Expr *typeCast = (Expr *) coerce_to_target_type(NULL,
+															(Node *) selectExpression,
+															selectOutputType,
+															columnType,
+															columnTypeMod,
+															COERCION_EXPLICIT,
+															COERCE_IMPLICIT_CAST,
+															-1);
+
+			selectTargetEntry->expr = typeCast;
+		}
+	}
+
+	multiPlan->insertSelectQuery = subquery;
```

Comment:
We modify the subquery, so the original INSERT..SELECT is no longer valid. Using it as a container for insertTargetList seems like it might cause bugs at some point.

There's also a possibility that we might use a different `targetRelationId` at some point. I'm thinking of cases in which there is an `ON CONFLICT` and under the covers we first copy into a temporary distributed table and then do a distributed `INSERT ... SELECT ... ON CONFLICT ...`.